File: native/android/app/src/main/kotlin/art/lazying/auspice/Localisation.kt
package art.lazying.auspice

import android.content.Context
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.unit.LayoutDirection
import java.util.Locale

/**
 * The interface language, and the one function that looks a string up.
 *
 * The app follows the phone unless the reader chooses otherwise in Settings,
 * and a choice takes effect at once rather than on the next launch.
 *
 */
object Localisation {
    private const val FILE = "auspice"
    private const val KEY = "language"
    private var context: Context? = null

    /** null means "follow the phone". */
    var chosen by mutableStateOf<String?>(null)
        private set

    fun load(appContext: Context) {
        context = appContext.applicationContext
        ReadingCatalogue.load(appContext)
        chosen = context?.getSharedPreferences(FILE, Context.MODE_PRIVATE)?.getString(KEY, null)
    }

    fun choose(code: String?) {
        chosen = code
        val editor = context?.getSharedPreferences(FILE, Context.MODE_PRIVATE)?.edit() ?: return
        if (code == null) editor.remove(KEY) else editor.putString(KEY, code)
        editor.apply()
    }

    val code: String
        get() = chosen?.takeIf { c -> Catalogue.languages.any { it.first == c } } ?: match()

    val layoutDirection: LayoutDirection
        get() = if (Catalogue.rightToLeft.contains(code)) LayoutDirection.Rtl else LayoutDirection.Ltr

    /**
     * Chinese needs care: a phone set to zh-TW or zh-HK wants the traditional
     * text, and everything else zh wants the simplified one.
     */
    private fun match(): String {
        val available = Catalogue.languages.map { it.first }
        val locale = Locale.getDefault()
        val tag = locale.toLanguageTag().lowercase()
        if (tag.startsWith("zh")) {
            val traditional = listOf("hant", "tw", "hk", "mo").any { tag.contains(it) }
            return if (traditional) "zh-Hant" else "zh-Hans"
        }
        available.firstOrNull { it.lowercase() == tag }?.let { return it }
        val base = tag.substringBefore('-')
        available.firstOrNull { it.lowercase() == base }?.let { return it }
        return "en"
    }
}

/** Look a string up in the language in force. */
fun t(key: String): String =
    if (key == "app.name" && BuildConfig.APPLICATION_ID == "art.lazying.lazyoracle") "LazyOracle" else
    Catalogue.table[key]?.get(Localisation.code)
        ?: Catalogue.table[key]?.get("en")
        ?: key

/** Presentation translations never alter engine identifiers or facts. */
object ReadingCatalogue {
    var values: Map<String, Map<String, String>> = emptyMap()
        private set
    fun load(context: Context) {
        values = context.assets.open("auspice-content.json").bufferedReader().use {
            kotlinx.serialization.json.Json.decodeFromString(it.readText())
        }
    }
}
fun l(source: String): String {
    ReadingCatalogue.values[source]?.get(Localisation.code)?.let { return it }
    Catalogue.glossary[source]?.get(Localisation.code)?.let { return it }
    for (separator in listOf(" · ", "、", " / ")) {
        if (source.contains(separator)) return source.split(separator).joinToString(" · ") { l(it) }
    }
    return source
}
fun lf(template: String, vararg values: String): String {
    var result = l(template)
    values.forEachIndexed { i, value -> result = result.replace("{$i}", l(value)) }
    return result
}
fun glossed(term: String): String = l(term)
fun lunarDateText(source: String): String {
    if (Localisation.code.startsWith("zh")) return if (Localisation.code == "zh-Hant") source.replace("闰", "閏") else source
    val digits = "〇一二三四五六七八九".withIndex().associate { it.value to it.index.toString() } + ('零' to "0")
    fun number(text: String): String {
        if (text == "正") return "1"
        if (text == "冬") return "11"
        if (text == "腊") return "12"
        val cleaned = text.replace("初", "").replace("廿", "二十").replace("卅", "三十")
        if (cleaned.contains("十")) {
            val parts = cleaned.split("十")
            return (((parts[0].firstOrNull()?.let { digits[it]?.toInt() } ?: 1) * 10) + (parts.last().firstOrNull()?.let { digits[it]?.toInt() } ?: 0)).toString()
        }
        return cleaned.map { digits[it] ?: it.toString() }.joinToString("")
    }
    val parts = source.split('年', '月', '日').filter { it.isNotEmpty() }
    if (parts.size !in 2..3) return l(source)
    val date = parts.joinToString("-") { number(it.replace("闰", "")) }
    return lf(if (source.contains("闰")) "Lunar date: {0} (leap month)" else "Lunar date: {0}", date)
}

fun readingLanguageInstruction(): String = "The interface language is ${Localisation.code}. Use it as the default, but naturally follow the language the reader uses or explicitly requests. Keep each answer coherent and easy to understand. Occasional useful terms from another language are fine; avoid unnecessary switching or duplicate translated paragraphs."

File: native/ios/Auspice/Localisation.swift
import SwiftUI

/// The interface language, and the one function that looks a string up.
///
/// The app follows the phone unless the reader chooses otherwise in Settings,
/// and a choice takes effect at once rather than on the next launch — which is
/// why the strings live in a table here rather than in `.lproj` bundles.
///
@Observable
final class Localisation {
    static let shared = Localisation()
    private static let key = "auspice.language"

    /// nil means "follow the phone".
    var chosen: String? {
        didSet {
            if let chosen {
                UserDefaults.standard.set(chosen, forKey: Self.key)
            } else {
                UserDefaults.standard.removeObject(forKey: Self.key)
            }
        }
    }

    private init() {
        chosen = UserDefaults.standard.string(forKey: Self.key)
    }

    /// The language actually in use: the chosen one, or the closest the phone
    /// asks for, or English.
    var code: String {
        if let chosen, Catalogue.table.values.first?[chosen] != nil { return chosen }
        return Self.match(Locale.preferredLanguages)
    }

    var layoutDirection: LayoutDirection {
        Catalogue.rightToLeft.contains(code) ? .rightToLeft : .leftToRight
    }

    /// Chinese needs care: a phone set to zh-TW or zh-HK wants the traditional
    /// text, and everything else zh wants the simplified one.
    static func match(_ preferred: [String]) -> String {
        let available = Catalogue.languages.map(\.code)
        for tag in preferred {
            let lower = tag.lowercased()
            if lower.hasPrefix("zh") {
                let traditional = ["hant", "tw", "hk", "mo"].contains { lower.contains($0) }
                return traditional ? "zh-Hant" : "zh-Hans"
            }
            if let exact = available.first(where: { $0.lowercased() == lower }) { return exact }
            let base = lower.split(separator: "-").first.map(String.init) ?? lower
            if let loose = available.first(where: { $0.lowercased() == base }) { return loose }
        }
        return "en"
    }
}

/// Look a string up in the language in force. A missing key shows itself
/// rather than an empty space, so a gap is obvious the first time it is seen.
func t(_ key: String) -> String {
    Catalogue.table[key]?[Localisation.shared.code] ?? Catalogue.table[key]?["en"] ?? key
}

/// Engine text is translated only for presentation. Identifiers and facts
/// sent back to the engines are never changed.
private enum ReadingCatalogue {
    static let values: [String: [String: String]] = {
        guard let url = Bundle.main.url(forResource: "auspice-content", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let values = try? JSONDecoder().decode([String: [String: String]].self, from: data) else { return [:] }
        return values
    }()
}

func l(_ source: String) -> String {
    let code = Localisation.shared.code
    if let translated = ReadingCatalogue.values[source]?[code] { return translated }
    if let translated = Catalogue.glossary[source]?[code] { return translated }
    // Lists of engine terms keep their order and receive no extra meanings.
    for separator in [" · ", "、", " / "] where source.contains(separator) {
        return source.components(separatedBy: separator).map(l).joined(separator: " · ")
    }
    return source
}

func lf(_ template: String, _ values: String...) -> String {
    var result = l(template)
    for (index, value) in values.enumerated() {
        result = result.replacingOccurrences(of: "{\(index)}", with: l(value))
    }
    return result
}

func glossed(_ term: String) -> String { l(term) }

func lunarDateText(_ source: String) -> String {
    if Localisation.shared.code.hasPrefix("zh") {
        return Localisation.shared.code == "zh-Hant" ? source.replacingOccurrences(of: "闰", with: "閏").replacingOccurrences(of: "腊", with: "臘") : source
    }
    let digits: [Character: String] = ["〇":"0", "零":"0", "一":"1", "二":"2", "三":"3", "四":"4", "五":"5", "六":"6", "七":"7", "八":"8", "九":"9"]
    func number(_ text: String) -> String {
        if text == "正" { return "1" }; if text == "冬" { return "11" }; if text == "腊" { return "12" }
        let cleaned = text.replacingOccurrences(of: "初", with: "").replacingOccurrences(of: "廿", with: "二十").replacingOccurrences(of: "卅", with: "三十")
        if cleaned.contains("十") {
            let parts = cleaned.components(separatedBy: "十")
            let tens = parts[0].first.flatMap { digits[$0] }.flatMap(Int.init) ?? 1
            let ones = parts.last?.first.flatMap { digits[$0] }.flatMap(Int.init) ?? 0
            return String(tens * 10 + ones)
        }
        return cleaned.map { digits[$0] ?? String($0) }.joined()
    }
    let parts = source.components(separatedBy: CharacterSet(charactersIn: "年月日")).filter { !$0.isEmpty }
    guard parts.count == 3 else { return l(source) }
    let leap = parts[1].contains("闰")
    let date = parts.enumerated().map { number($0.element.replacingOccurrences(of: "闰", with: "")) }.joined(separator: "-")
    return lf(leap ? "Lunar date: {0} (leap month)" : "Lunar date: {0}", date)
}

/// The selected language leads; an occasional traditional term is fine.
func readingLanguageInstruction() -> String {
    "Write mainly in the selected app language: \(Localisation.shared.code). Use clear everyday language. Traditional Chinese uses Traditional characters; Simplified Chinese uses Simplified characters. Brief conventional terms from another language are allowed when useful, but do not alternate languages, duplicate paragraphs in translation, or follow the language of raw engine data. Explain unfamiliar terms once. If the reader explicitly requests a different response language, honor that request."
}
